Core/MIPS/x86/X64IRCompSystem.cpp

```diff
@@ -201,21 +201,90 @@ void X64JitBackend::CompIR_Transfer(IRInst inst) {
 	}
 }
 
+int ReportBadAddress(uint32_t addr, uint32_t alignment, uint32_t isWrite) {
+	const auto toss = [&](MemoryExceptionType t) {
+		Core_MemoryException(addr, alignment, currentMIPS->pc, t);
+		return coreState != CORE_RUNNING ? 1 : 0;
+	};
+
+	if (!Memory::IsValidRange(addr, alignment)) {
+		MemoryExceptionType t = isWrite == 1 ? MemoryExceptionType::WRITE_WORD : MemoryExceptionType::READ_WORD;
+		if (alignment > 4)
+			t = isWrite ? MemoryExceptionType::WRITE_BLOCK : MemoryExceptionType::READ_BLOCK;
+		return toss(t);
+	} else if (alignment > 1 && (addr & (alignment - 1)) != 0) {
+		return toss(MemoryExceptionType::ALIGNMENT);
+	}
+	return 0;
+};
+
 void X64JitBackend::CompIR_ValidateAddress(IRInst inst) {
 	CONDITIONAL_DISABLE;
 
+	bool isWrite = inst.src2 & 1;
+	int alignment = 0;
 	switch (inst.op) {
 	case IROp::ValidateAddress8:
+		alignment = 1;
+		break;
+
 	case IROp::ValidateAddress16:
+		alignment = 2;
+		break;
+
 	case IROp::ValidateAddress32:
+		alignment = 4;
+		break;
+
 	case IROp::ValidateAddress128:
-		CompIR_Generic(inst);
+		alignment = 16;
 		break;
 
 	default:
 		INVALIDOP;
 		break;
 	}
+
+	// This is unfortunate...
+	FlushAll();
```

Comment:
hm, instead of flushing, could we call a thunk to save the registers around the ReportBadAddress call?

Reply:
Well, it needs to bail if it fails, so not a thunk.  What we do in the other jit is save the state of the regcache and then restore it, and do the flush only then.  But that's actually terrible on codesize.  We do use a func for the actual checks, though, which probably helps branch prediction.

-[Unknown]

Reply:
Oh right. Although, we could also save off some structure elsewhere with information about the allocated registers for each case like this, and compile some function that use that information to writeback the registers according to that. Though complex.